File: iosApp/iosApp/core/model/WallpaperUi.swift
//
//  WallpaperUi.swift
//  iosApp
//
//  Created by platinum on 26/12/2025.
//

import Foundation
import Shared
import SwiftUI

struct WallpaperUi: Identifiable, Equatable, Hashable {
    let id: Int64
    let imageUrl: String
    let smallImageUrl: String
    let photographerName: String
    let photographerUrl: String
    let averageColor: String
    let height: Int
    let width: Int
    var isFavorite: Bool
    let addedAt: Int64
}

extension WallpaperUi {
    func copy(isFavorite: Bool? = nil) -> WallpaperUi {
        return WallpaperUi(
            id: self.id,
            imageUrl: self.imageUrl,
            smallImageUrl: self.smallImageUrl,
            photographerName: self.photographerName,
            photographerUrl: self.photographerUrl,
            averageColor: self.averageColor,
            height: self.height,
            width: self.width,
            isFavorite: isFavorite ?? self.isFavorite,
            addedAt: addedAt
        )
    }

    func toDomain() -> Wallpaper {
        return Wallpaper(
            id: self.id,
            imageUrl: self.imageUrl,
            smallImageUrl: self.smallImageUrl,
            photographer: self.photographerName,
            photographerUrl: self.photographerUrl,
            averageColor: self.averageColor,
            height: Int32(self.height),
            width: Int32(self.width),
            isFavorite: isFavorite,
            addedAt: addedAt
        )
    }
}

extension Wallpaper {
    func toUi() -> WallpaperUi {
        return WallpaperUi(
            id: self.id,
            imageUrl: self.imageUrl,
            smallImageUrl: self.smallImageUrl,
            photographerName: self.photographer,
            photographerUrl: self.photographerUrl,
            averageColor: self.averageColor,
            height: Int(self.height),
            width: Int(self.width),
            isFavorite: self.isFavorite,
            addedAt: addedAt
        )
    }
}

File: app/src/main/kotlin/com/example/aura/shared/model/WallpaperUi.kt
package com.example.aura.shared.model

import androidx.compose.runtime.Immutable
import com.example.aura.domain.model.Wallpaper
import kotlinx.serialization.Serializable


@Immutable
@Serializable
data class WallpaperUi(
    val id: Long,
    val imageUrl: String,
    val smallImageUrl: String,
    val photographerName: String,
    val contentDescription: String,
    val aspectRatio: Float,
    val averageColor: String,
    val isFavorite: Boolean
)

fun Wallpaper.toUi(): WallpaperUi {
    return WallpaperUi(
        id = id,
        imageUrl = imageUrl,
        smallImageUrl = smallImageUrl,
        photographerName = "Photo by ${photographer}",
        contentDescription = "Photo by ${photographer}",
        aspectRatio = if (height > 0) width.toFloat() / height.toFloat() else 0.7f,
        averageColor = averageColor,
        isFavorite = isFavorite
    )
}